tools/lxdclient/client_image.go

```diff
@@ -92,56 +97,107 @@ func (p *progressContext) copyProgress(progress string) {
 			lastErr = err
 			continue
 		}
-
-		// TODO(jam): there are multiple possible spellings for aliases,
-		// unfortunately. cloud-images only hosts ubuntu images, and
-		// aliases them as "trusty" or "trusty/amd64" or
-		// "trusty/amd64/20160304". However, we should be more
-		// explicit. and use "ubuntu/trusty/amd64" as our default
-		// naming scheme, and only fall back for synchronization.
-		target := source.GetAlias(series)
-		if target == "" {
-			logger.Infof("no image for %s found in %s", name, source.URL())
-			// TODO(jam) Add a test that we skip sources that don't
-			// have what we are looking for
+		err = i.ensureImage(series, imageName, aliases, source, copyProgressHandler)
+		if errors.IsNotFound(err) {
 			continue
 		}
-		logger.Infof("found image from %s for %s = %s",
-			source.URL(), series, target)
-		forwarder := stringforwarder.New(copyProgressHandler)
-		defer func() {
-			dropCount := forwarder.Stop()
-			logger.Debugf("dropped %d progress messages", dropCount)
-		}()
-		adapter := &progressContext{
-			logger:  logger,
-			level:   loggo.INFO,
-			context: fmt.Sprintf("copying image for %s from %s: %%s", name, source.URL()),
-			forward: forwarder.Forward,
+		if lastErr = err; lastErr == nil {
+			break
+		}
+	}
+	return imageName, lastErr
+}
+
+func (i *imageClient) ensureImage(
+	series, imageName string,
+	aliases []string,
+	source remoteClient,
+	copyProgressHandler func(string),
+) error {
+	// Look for an image with any of the aliases.
+	var alias, target string
+	for _, alias = range aliases {
+		if target = source.GetAlias(alias); target != "" {
+			break
 		}
-		err = source.CopyImage(series, i.raw, []string{name}, adapter.copyProgress)
-		return errors.Annotatef(err, "unable to get LXD image for %s", name)
 	}
-	return lastErr
+	if target == "" {
+		// TODO(jam) Add a test that we skip sources that don't
+		// have what we are looking for
+		logger.Infof("no image for %s found in %s", imageName, source.URL())
+		return errors.NotFoundf("image for %s in %s", imageName, source.URL())
+	}
+
+	logger.Infof("found image from %s for %s = %s", source.URL(), imageName, target)
+	forwarder := stringforwarder.New(copyProgressHandler)
+	defer func() {
+		dropCount := forwarder.Stop()
+		logger.Debugf("dropped %d progress messages", dropCount)
+	}()
+	adapter := &progressContext{
+		logger:  logger,
+		level:   loggo.INFO,
+		context: fmt.Sprintf("copying image for %s from %s: %%s", imageName, source.URL()),
+		forward: forwarder.Forward,
+	}
+	err := source.CopyImage(alias, i.raw, []string{imageName}, adapter.copyProgress)
+	return errors.Annotatef(err, "unable to get LXD image for %s", imageName)
+}
+
+// seriesLocalAlias returns the alias to assign to images for the
+// specified series. The alias is juju-specific, to support the
+// user supplying a customised image (e.g. CentOS with cloud-init).
+func seriesLocalAlias(series, arch string) string {
+	return fmt.Sprintf("juju/%s/%s", series, arch)
```

Comment:
FWIW, I prefer the foo/bar/baz path, because that does seem to be what upstream has been using. I was going to switch to it for "ubuntu/trusty" but it didn't seem worth it to break compatibility at the time.
If we're going with the full names, then we'll want to do this.